kgtk/graph_embeddings/build_faiss.py

```diff
@@ -0,0 +1,181 @@
+import numpy as np
+import faiss
+import os
+from tqdm import tqdm
+from kgtk.exceptions import KGTKException
+
+
+# # TODO:
+# - look up what is faiss.set_direct_map_type and if I need more params for this.
+#   - still not sure. Very little information about it is easily available
+# - add support for sharding
+# - add support for cosine similarity as a distance metric
+
+
+def build_faiss(embeddings_file, embeddings_format, no_input_header, index_file_out, index_to_node_file_out,
+                max_train_examples, workers, index_string, metric_type, p=None, verbose=False):
+
+    # validate input file path
+    if not os.path.exists(embeddings_file):
+        raise KGTKException("File path given for embeddings_file parameter does not exist: " + embeddings_file)
+    if os.path.getsize(embeddings_file) == 0:
+        raise KGTKException("File given for embeddings_file parameter is empty: " + embeddings_file)
+
+    # validate metric type and translate to a faiss metric
+    metrics = {
+        "Inner_product": faiss.METRIC_INNER_PRODUCT,
+        "L2": faiss.METRIC_L2,
+        "L1": faiss.METRIC_L1,
+        "Linf": faiss.METRIC_Linf,
+        "Lp": faiss.METRIC_Lp,
+        "Canberra": faiss.METRIC_Canberra,
+        "BrayCurtis": faiss.METRIC_BrayCurtis,
+        "JensenShannon": faiss.METRIC_JensenShannon
+    }
+    if metric_type in metrics:
+        faiss_metric = metrics[metric_type]
+    else:
+        raise KGTKException("Unrecognized value for metric_type parameter: {}.".format(metric_type) +
+                            "Please choose one of {}.".format(" | ".join(list(metrics.keys()))))
+    if metric_type == "Lp" and p is None:
```

Comment:
what is `p`? Use a descriptive name please

Reply:
If the Lp distance metric is specified, then the user also needs to specify the value of p that they want to use, e.g. L1, L2, L...
I'll change it to metric_arg though since that is what Faiss calls it.